filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,519 @@
+	reuseDuration time.Duration
+}
+
+func WithReuseDuration(duration time.Duration) func(*OpenPolicyAgentRegistry) error {
+	return func(cfg *OpenPolicyAgentRegistry) error {
+		cfg.reuseDuration = duration
+		return nil
+	}
+}
+
+const defaultReuseDuration = 30 * time.Second
+const defaultShutdownGracePeriod = 30 * time.Second
+
+func NewOpenPolicyAgentRegistry(opts ...func(*OpenPolicyAgentRegistry) error) *OpenPolicyAgentRegistry {
+	registry := &OpenPolicyAgentRegistry{
+		reuseDuration: defaultReuseDuration,
+		instances:     make(map[string]*OpenPolicyAgentInstance),
+		refcounts:     make(map[*OpenPolicyAgentInstance]int),
+		lastused:      make(map[*OpenPolicyAgentInstance]time.Time),
+		quit:          make(chan struct{}),
+	}
+
+	for _, opt := range opts {
+		opt(registry)
+	}
+
+	go registry.startCleanerDaemon()
+
+	return registry
+}
+
+type OpenPolicyAgentInstanceConfig struct {
+	envoyMetadata  *ext_authz_v3_core.Metadata
+	configTemplate []byte
+}
+
+func WithConfigTemplate(configTemplate []byte) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.configTemplate = configTemplate
+		return nil
+	}
+}
+
+func WithConfigTemplateFile(configTemplateFile string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		var err error
+		cfg.configTemplate, err = os.ReadFile(configTemplateFile)
+		return err
+	}
+}
+
+func WithEnvoyMetadata(metadata *ext_authz_v3_core.Metadata) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyMetadata = metadata
+		return nil
+	}
+}
+
+func WithEnvoyMetadataBytes(content []byte) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyMetadata = &ext_authz_v3_core.Metadata{}
+
+		return protojson.Unmarshal(content, cfg.envoyMetadata)
+	}
+}
+
+func WithEnvoyMetadataFile(file string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		content, err := os.ReadFile(file)
+		if err != nil {
+			return err
+		}
+
+		err = WithEnvoyMetadataBytes(content)(cfg)
+		if err != nil {
+			return fmt.Errorf("cannot parse '%v': %w", file, err)
+		}
+
+		return nil
+	}
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetEnvoyMetadata() *ext_authz_v3_core.Metadata {
+	return cfg.envoyMetadata
+}
+
+func NewOpenPolicyAgentConfig(opts ...func(*OpenPolicyAgentInstanceConfig) error) (*OpenPolicyAgentInstanceConfig, error) {
+	cfg := OpenPolicyAgentInstanceConfig{}
+
+	for _, opt := range opts {
+		if err := opt(&cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	if cfg.configTemplate == nil {
+		var err error
+		cfg.configTemplate, err = os.ReadFile(configTemplate())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &cfg, nil
+}
+
+func configTemplate() string {
+	if s, ok := os.LookupEnv("OPA_CONFIG_TEMPLATE_FILE"); ok {
```

Comment:
I thought we have `-open-policy-agent-config-template` flag.
I think generally we avoid using env variables.

Reply:
Removed it, this is from before moving this here.